b2sum.1
.TH B2SUM 1 BLAKESUM
.SH NAME
b2sum - Compute and check BLAKE2 message digests
.SH SYNOPSIS
.B b2sum
[-l bits | -X bits] [-K key] [-P pepper] [-S salt] [-c | -B | -L | -U] [-sxz]
.RI [ file "] ..."
.SH DESCRIPTION
Print or check BLAKE2 checksums, using BLAKE2b by default.
.SH OPTIONS
The
.B b2sum
utility conforms to the Base Definitions volume of POSIX.1-2017,
.IR "Section 12.2" ,
.IR "Utility Syntax Guidelines" .
.PP
The following options are supported:
.TP
.B -B
Output checksums in binary representation. This suppresses
the filenames and checksum delimiters. Only the checksums
are printed.
.TP
.B -c
Read BLAKE2 sums from the file and check them against the
files on your system. The input files should be
formatted as the output of this program, or similarly.
This is not going to work if any of the filenames in the
input files starts with <space> or <tab>, or if they
contain a <newline>, unless the
.B -z
option is also used.
.TP
.BI "-K " key
Specify a key (which is used for MAC and PRF), that is
up to 32 bytes (if the
.B -s
option is used) or 64 bytes (otherwise) long, that the
algorithm shall use. The key shall be expressed in
hexadecimal: each byte in the key shall be expressed, from
left to right, as a pairs of hexadecimal digits where highest
bits in each byte is stored in the left digit in its pair and
the lowest bits in each byte is stored in the right digit in
its pair. No delimiters are used, so for a 4 byte key where
each byte's value is its index, the key is expessed as
.BR 00010203 ,
however each letter may be either small or capital.
The key may not be empty.
.TP
.B -L
Output checksums in lower-case hexadecimal representation. (Default)
.TP
.BI "-l " bits
Select output size, valid values are multiples of 8
between 8 and 512, inclusively, or if the
.B -s
option is used between 8 and 256, inclusively.
(Default is maximum.)
.TP
.BI "-P " pepper
Specify an 8-byte (if the
.B -s
option is used) or 16-byte pepper (\(dqpersonalisation\(dq)
(otherwise) that the algorithm shall use. This pepper shall
be expressed in full length hexadecimal: 16 (for 8-byte) or 32
(for 16-byte) hexadecimal digits, or rather 8 or 16 pairs
of hexadecimal digits, ordered from left to right to specify
the values from index 0 to the last byte. In each pair,
the left digit stores the high bits of the byte, and the
right digit stores the low bits of the byte. For example,
the digit pair
.B 80
represents a byte with the decimal value 128, whereas the
digit pair
.B 08
represents a byte with the decimal value 8. The pairs
are joined without any delimiters, and no byte may be
omitted. So, for a 8 byte pepper where each byte's value
is its index, the pepper is expressed as
.BR 0001020304050607 ,
however each letter may be either small or capital.
If no pepper is specified, an all-zeroes pepper is used.
.TP
.BI "-S " salt
Specify a 8-byte (if the
.B -s
option is used) or 16-byte salt
(otherwise) that the algorithm shall use. This salt shall
be expressed in full length hexadecimal: 16 (for 8-byte) or 32
(for 16-byte) hexadecimal digits, or rather 8 or 16 pairs
of hexadecimal digits, ordered from left to right to specify
the values from index 0 to the last byte. In each pair,
the left digit stores the high bits of the byte, and the
right digit stores the low bits of the byte. For example,
the digit pair
.B 80
represents a byte with the decimal value 128, whereas the
digit pair
.B 08
represents a byte with the decimal value 8. The pairs
are joined without any delimiters, and no byte may be
omitted. So, for a 8 byte salt where each byte's value
is its index, the salt is expressed as
.BR 0001020304050607 ,
however each letter may be either small or capital.
If no salt is specified, an all-zeroes salt is used.
.TP
.B -s
Use BLAKE2s instead of BLAKE2b.
.TP
.B -U
Output checksums in upper-case hexadecimal representation.
.TP
.BI "-X " bits
Use BLAKE2Xb instead of BLAKE2b or BLAKE2Xs instead of BLAKE2s,
and select the extended output function output length, in bits;
must be a non-zero multiple of 8, and no greater than
34359738360 or (if the
.B -s
option is used) 524280.
.TP
.B -x
Convert input files from hexadecimal form to binary form
before calculating the checksums.
.TP
.B -z
Lines end with NUL instead of LF. If used with
.BR -c ,
this applies to read files (not the output), but it will
also apply more strict parsing and allow any whitespace
in file names.
.SH OPERANDS
The following operands are supported:
.TP
.I file
File to read. The standard input will be used
.B -
or no
.I file
is specified.
.SH EXIT STATUS
.TP
0
Successful completion.
.TP
1
Checksums did not match or a file did not exist.
.TP
2
An error occurred.
.SH NOTES
BLAKE2X has not been finalised as of 2022-02-20.
.SH SEE ALSO
.BR bsum (1)


b2sum.c
/* See LICENSE file for copyright and license details. */
#include "common.h"

const char *argv0 = "b2sum";

static int flag_check = 0;
static int flag_binary = 0;
static int flag_lower = 0;
static int flag_upper = 0;
static int flag_small = 0;
static int flag_extended = 0;
static int flag_hex = 0;
static int flag_zero = 0;
static int length = 0;
static long long int xlength = 0;
static void *key = NULL;
static size_t key_len = 0;
static void *salt = NULL;
static void *pepper = NULL;

static size_t hashlen;

static void
usage(void)
{
	/* TODO add support for parallel versions */
	/* TODO add support for tree hashing */
	fprintf(stderr, "usage: %s [-l bits | -X bits] [-K key] [-P pepper] [-S salt]"
	                " [-c | -B | -L | -U] [-sxz] [file] ...\n", argv0);
	exit(2);
}

static int
hash_fd_blake2bs(int fd, const char *fname, int decode_hex, unsigned char hash[])
{
	struct libblake_blake2b_state state2b;
	struct libblake_blake2b_params params2b;
	struct libblake_blake2s_state state2s;
	struct libblake_blake2s_params params2s;
	char *buf = NULL;
	size_t size = 0;
	size_t len = 0;
	size_t off = 0;
	size_t req;
	ssize_t r;
	int ok;
	if (flag_small) {
		memset(&params2s, 0, sizeof(params2s));
		params2s.digest_len = (uint_least8_t)length;
		params2s.key_len = (uint_least8_t)key_len;
		params2s.fanout = 1;
		params2s.depth = 1;
		if (salt)
			memcpy(params2s.salt, salt, sizeof(params2s.salt));
		if (pepper)
			memcpy(params2s.pepper, pepper, sizeof(params2s.pepper));
		libblake_blake2s_init(&state2s, &params2s);
		if (key) {
			buf = erealloc(buf, size = 8 << 10);
			len = 64;
			memcpy(buf, key, len);
			off += libblake_blake2s_update(&state2s, key, len);
		}
	} else {
		memset(&params2b, 0, sizeof(params2b));
		params2b.digest_len = (uint_least8_t)length;
		params2b.key_len = (uint_least8_t)key_len;
		params2b.fanout = 1;
		params2b.depth = 1;
		if (salt)
			memcpy(params2b.salt, salt, sizeof(params2b.salt));
		if (pepper)
			memcpy(params2b.pepper, pepper, sizeof(params2b.pepper));
		libblake_blake2b_init(&state2b, &params2b);
		if (key) {
			buf = erealloc(buf, size = 8 << 10);
			len = 128;
			memcpy(buf, key, len);
			off += libblake_blake2b_update(&state2b, key, len);
		}
	}
	for (;;) {
		if (len == size)
			buf = erealloc(buf, size += 8 << 10);
		r = read(fd, &buf[len], size - len);
		if (r <= 0) {
			if (!r)
				break;
			if (errno == EINTR)
				continue;
			fprintf(stderr, "%s: %s: %s\n", argv0, fname, strerror(errno));
			free(buf);
			return -1;
		}
		len += (size_t)r;
		if (!decode_hex) {
			if (flag_small)
				off += libblake_blake2s_update(&state2s, &buf[off], len - off);
			else
				off += libblake_blake2b_update(&state2b, &buf[off], len - off);
			if (off == len)
				off = 0;
		}
	}
	if (off)
		memmove(&buf[0], &buf[off], len -= off);
	if (decode_hex) {
		len = libblake_decode_hex(buf, len, buf, &ok);
		if (!ok) {
			fprintf(stderr, "%s: %s: %s\n", argv0, fname, "invalid hexadecimal input");
			free(buf);
			return -1;
		}
	}
	if (flag_small)
		req = libblake_blake2s_digest_get_required_input_size(len);
	else
		req = libblake_blake2b_digest_get_required_input_size(len);
	if (req > size)
		buf = erealloc(buf, size);
	if (flag_small)
		libblake_blake2s_digest(&state2s, buf, len, 0, hashlen / 8, hash);
	else
		libblake_blake2b_digest(&state2b, buf, len, 0, hashlen / 8, hash);
	free(buf);
	return 0;
}

static int
hash_fd_blake2xbs(int fd, const char *fname, int decode_hex, unsigned char hash[])
{
	struct libblake_blake2xb_state state2xb;
	struct libblake_blake2xb_params params2xb;
	struct libblake_blake2xs_state state2xs;
	struct libblake_blake2xs_params params2xs;
	char *buf = NULL;
	size_t size = 0;
	size_t len = 0;
	size_t off = 0;
	size_t req;
	size_t i, n;
	ssize_t r;
	int ok;
	if (flag_small) {
		memset(&params2xs, 0, sizeof(params2xs));
		params2xs.digest_len = (uint_least8_t)length;
		params2xs.key_len = (uint_least8_t)key_len;
		params2xs.fanout = 1;
		params2xs.depth = 1;
		params2xs.xof_len = (uint_least16_t)xlength;
		if (salt)
			memcpy(params2xs.salt, salt, sizeof(params2xs.salt));
		if (pepper)
			memcpy(params2xs.pepper, pepper, sizeof(params2xs.pepper));
		libblake_blake2xs_init(&state2xs, &params2xs);
		if (key) {
			buf = erealloc(buf, size = 8 << 10);
			len = 64;
			memcpy(buf, key, len);
			off += libblake_blake2xs_update(&state2xs, key, len);
		}
	} else {
		memset(&params2xb, 0, sizeof(params2xb));
		params2xb.digest_len = (uint_least8_t)length;
		params2xb.key_len = (uint_least8_t)key_len;
		params2xb.fanout = 1;
		params2xb.depth = 1;
		params2xb.xof_len = (uint_least32_t)xlength;
		if (salt)
			memcpy(params2xb.salt, salt, sizeof(params2xb.salt));
		if (pepper)
			memcpy(params2xb.pepper, pepper, sizeof(params2xb.pepper));
		libblake_blake2xb_init(&state2xb, &params2xb);
		if (key) {
			buf = erealloc(buf, size = 8 << 10);
			len = 128;
			memcpy(buf, key, len);
			off += libblake_blake2xb_update(&state2xb, key, len);
		}
	}
	for (;;) {
		if (len == size)
			buf = erealloc(buf, size += 8 << 10);
		r = read(fd, &buf[len], size - len);
		if (r <= 0) {
			if (!r)
				break;
			if (errno == EINTR)
				continue;
			fprintf(stderr, "%s: %s: %s\n", argv0, fname, strerror(errno));
			free(buf);
			return -1;
		}
		len += (size_t)r;
		if (!decode_hex) {
			if (flag_small)
				off += libblake_blake2xs_update(&state2xs, &buf[off], len - off);
			else
				off += libblake_blake2xb_update(&state2xb, &buf[off], len - off);
			if (off == len)
				off = 0;
		}
	}
	if (off)
		memmove(&buf[0], &buf[off], len -= off);
	if (decode_hex) {
		len = libblake_decode_hex(buf, len, buf, &ok);
		if (!ok) {
			fprintf(stderr, "%s: %s: %s\n", argv0, fname, "invalid hexadecimal input");
			free(buf);
			return -1;
		}
	}
	if (flag_small)
		req = libblake_blake2xs_predigest_get_required_input_size(&state2xs);
	else
		req = libblake_blake2xb_predigest_get_required_input_size(&state2xb);
	if (req > size)
		buf = erealloc(buf, size);
	if (flag_small)
		libblake_blake2xs_predigest(&state2xs, buf, len, 0);
	else
		libblake_blake2xb_predigest(&state2xb, buf, len, 0);
	if (flag_small) {
		for (i = 0; i * 32 < hashlen / 8; i++) { /* TODO this could be done parallel (but align hash) (also below) */
			n = (i + 1) * 32 > hashlen / 8 ? hashlen / 8 - i * 32 : 32;
			libblake_blake2xs_digest(&state2xs, (uint_least32_t)i, (uint_least8_t)n, &hash[i * 32]);
		}
	} else {
		for (i = 0; i * 64 < hashlen / 8; i++) {
			n = (i + 1) * 64 > hashlen / 8 ? hashlen / 8 - i * 64 : 64;
			libblake_blake2xb_digest(&state2xb, (uint_least32_t)i, (uint_least8_t)n, &hash[i * 64]);
		}
	}
	free(buf);
	return 0;
}

int
hash_fd(int fd, const char *fname, int decode_hex, unsigned char hash[])
{
	int ret;

	if (flag_extended)
		ret = hash_fd_blake2xbs(fd, fname, decode_hex, hash);
	else
		ret = hash_fd_blake2bs(fd, fname, decode_hex, hash);

	return ret;
}

int
main(int argc, char *argv[])
{
	const char *key_str = NULL;
	uint_least8_t key_buf[128];
	const char *salt_str = NULL;
	uint_least8_t salt_buf[16];
	const char *pepper_str = NULL;
	uint_least8_t pepper_buf[16];
	int status = 0;
	int output_case;
	char newline;

	ARGBEGIN {
	case 'c':
		flag_check = 1;
		break;
	case 'B':
		flag_binary = 1;
		break;
	case 'L':
		flag_lower = 1;
		flag_upper = 0;
		break;
	case 'U':
		flag_upper = 1;
		flag_lower = 0;
		break;
	case 'K':
		if (key_str)
			usage();
		key_str = ARG();
		break;
	case 'S':
		if (salt_str)
			usage();
		salt_str = ARG();
		break;
	case 'P':
		if (pepper_str)
			usage();
		pepper_str = ARG();
		break;
	case 's':
		flag_small = 1;
		break;
	case 'x':
		flag_hex = 1;
		break;
	case 'z':
		flag_zero = 1;
		break;
	case 'l':
		if (length)
			usage();
		length = atoi(ARG());
		if (length & 7 || !length || length > 512) {
			fprintf(stderr, "%s: valid arguments for -l\n", argv0);
			return 2;
		}
		break;
	case 'X':
		if (flag_extended)
			usage();
		flag_extended = 1;
		xlength = atoll(ARG());
		if (xlength & 7 || !xlength || xlength > 34359738360LL) {
			fprintf(stderr, "%s: valid arguments for -X\n", argv0);
			return 2;
		}
		break;
	default:
		usage();
	} ARGEND;

	if (flag_check + flag_binary + flag_lower + flag_upper > 1 ||
	    (length && flag_extended))
		usage();

	if (!length)
		length = flag_small ? 256 : 512;
	else if (flag_small && length > 256)
		fprintf(stderr, "%s: valid arguments for -l\n", argv0);
	else if (flag_small && xlength > 524280LL)
		fprintf(stderr, "%s: valid arguments for -X\n", argv0);

	if (key_str) {
		memset(key_buf, 0, sizeof(key_buf));
		key_len = parse_key(key_buf, key_str, flag_small ? 32 : 64);
		key = key_buf;
	}

	if (pepper_str) {
		parse_pepper(pepper_buf, pepper_str, flag_small ? 8 : 16);
		pepper = pepper_buf;
	}

	if (salt_str) {
		parse_salt(salt_buf, salt_str, flag_small ? 8 : 16);
		salt = salt_buf;
	}

	hashlen = flag_extended ? (size_t)xlength : (size_t)length;
	length /= 8;
	xlength /= 8;

	newline = flag_zero ? '\0' : '\n';
	if (flag_check) {
		if (!argc) {
			status |= -check_and_print("-", hashlen, flag_hex, newline);
		} else {
			for (; *argv; argv++)
				status |= -check_and_print(*argv, hashlen, flag_hex, newline);
		}
	} else {
		output_case = flag_binary ? -1 : flag_upper;
		if (!argc) {
			status |= -hash_and_print("-", hashlen, flag_hex, newline, output_case);
		} else {
			for (; *argv; argv++)
				status |= -hash_and_print(*argv, hashlen, flag_hex, newline, output_case);
		}
	}

	if (fflush(stdout) || ferror(stdout) || fclose(stdout)) {
		fprintf(stderr, "%s: %s\n", argv0, strerror(errno));
		return 2;
	}
	return status;
}
